src/framework/constraints.ml

```diff
@@ -1694,6 +1694,162 @@ struct
+
+  (* Mapping from the callee context to the set of all caller tuples*)
+  module CallGraphMap = MapDomain.MapBot (S.C) (CallGraphSet)
+
+  module G =
+  struct
+    include Lattice.Lift2 (G) (CallGraphMap) (Printable.DefaultNames)
+
+    let spec = function
+      | `Bot -> G.bot ()
+      | `Lifted1 x -> x
+      | _ -> failwith "RecursionTermLifter.spec"
+    let callGraph = function
+      | `Bot -> CallGraphMap.bot ()
+      | `Lifted2 x -> x
+      | _ -> failwith "RecursionTermLifter.callGraph"
+    let create_spec spec = `Lifted1 spec
+    let create_callGraph callGraph = `Lifted2 callGraph
+
+    let printXml f = function
+      | `Lifted1 x -> G.printXml f x
+      | `Lifted2 x -> BatPrintf.fprintf f "<analysis name=\"recTerm-context\">%a</analysis>" CallGraphMap.printXml x
+      | x -> BatPrintf.fprintf f "<analysis name=\"recTerm\">%a</analysis>" printXml x
+
+  end
+
+  let name () = "RecursionTermLifter (" ^ S.name () ^ ")"
+
+  let conv (ctx: (_, G.t, _, V.t) ctx): (_, S.G.t, _, S.V.t) ctx =
+    { ctx with
+      global = (fun v -> G.spec (ctx.global (V.spec v)));
+      sideg = (fun v g -> ctx.sideg (V.spec v) (G.create_spec g));
+    }
+
+  let cycleDetection ctx v v' =
+    let module LH = Hashtbl.Make (Printable.Prod (CilType.Fundec) (S.C)) in
+    let module LS = Set.Make (Printable.Prod (CilType.Fundec) (S.C)) in
+    (* find all cycles/SCCs *)
+    let global_visited_calls = LH.create 100 in
+
+    (* DFS *)
+    let rec iter_call (path_visited_calls: LS.t) (call:Printable.Prod (CilType.Fundec) (S.C).t) =
+      let ((fundec_e:fundec), (context_e: C.t)) = call in (*unpack tuple for later use*)
+      if LS.mem call path_visited_calls then (
+        AnalysisState.svcomp_may_not_terminate := true; (*set the indicator for a non-terminating program for the sv comp*)
+        (*Cycle found*)
+        let msgs =
+          [
+            (Pretty.dprintf "The program might not terminate! (Fundec %a is contained in a call graph cycle)" CilType.Fundec.pretty fundec_e, Some (M.Location.CilLocation fundec_e.svar.vdecl));
+          ] in
+        M.msg_group Warning ~category:NonTerminating "Recursion cycle" msgs) (* output a warning for non-termination*)
+      else if not (LH.mem global_visited_calls call) then begin
+        LH.replace global_visited_calls call ();
+        let new_path_visited_calls = LS.add call path_visited_calls in
+        let fundec_e_typeV: V.t = V.relift (`Right fundec_e) in
+        let gmap = G.callGraph (ctx.global (fundec_e_typeV)) in
+        let callers: CallGraphSet.t = CallGraphMap.find (context_e) gmap in
+        CallGraphSet.iter (fun to_call ->
+            iter_call new_path_visited_calls to_call
+          ) callers;
+      end
+    in
+    let gmap = G.callGraph (ctx.global (v)) in
+    CallGraphMap.iter(fun key value ->
+        let call = (v', key) in
+        iter_call LS.empty call
+      ) gmap (* try all fundec + context pairs that are in the map *)
```

Comment:
The `G.callGraph`, ... `CallGraphSet.iter` logic is duplicated here. If the entry point of the recursive function were slightly different, this duplication could maybe be avoided.

Deadlock analysis does something similar but doesn't have such duplication.

Reply:
Fixed.